tests/integration/ha/ha_helpers.py

```diff
@@ -25,44 +35,50 @@ class ProcessRunningError(Exception):
 
 
-async def get_topic_leader(ops_test: OpsTest, topic: str) -> int:
+async def get_topic_description(ops_test: OpsTest, topic: str) -> TopicDescription:
     """Get the broker with the topic leader.
 
     Args:
         ops_test: OpsTest utility class
         topic: the desired topic to check
     """
-    bootstrap_server = (
-        await get_address(ops_test=ops_test)
-        + f":{SECURITY_PROTOCOL_PORTS['SASL_PLAINTEXT'].client}"
-    )
+    bootstrap_servers = []
+    for unit in ops_test.model.applications[APP_NAME].units:
+        bootstrap_servers.append(
+            await get_address(ops_test=ops_test, unit_num=unit.name.split("/")[-1])
+            + f":{SECURITY_PROTOCOL_PORTS['SASL_PLAINTEXT'].client}"
+        )
 
-    result = check_output(
-        f"JUJU_MODEL={ops_test.model_full_name} juju ssh kafka/0 sudo -i 'charmed-kafka.topics --bootstrap-server {bootstrap_server} --command-config {KafkaSnap.CONF_PATH}/client.properties --describe --topic {topic}'",
+    output = check_output(
+        f"JUJU_MODEL={ops_test.model_full_name} juju ssh kafka/0 sudo -i 'charmed-kafka.topics --bootstrap-server {','.join(bootstrap_servers)} --command-config {KafkaSnap.CONF_PATH}/client.properties --describe --topic {topic}'",
         stderr=PIPE,
         shell=True,
         universal_newlines=True,
     )
+    result = TopicDescription()
+    result.leader = int(re.search(r"Leader: (\d+)", output)[1])
+    result.in_sync_replicas = {int(i) for i in re.search(r"Isr: ([\d,]+)", output)[1].split(",")}
 
-    return re.search(r"Leader: (\d+)", result)[1]
+    return result
 
 
-async def get_topic_offsets(ops_test: OpsTest, topic: str, unit_name: str) -> list[str]:
+async def get_topic_offsets(ops_test: OpsTest, topic: str) -> list[str]:
     """Get the offsets of a topic on a unit.
 
     Args:
         ops_test: OpsTest utility class
         topic: the desired topic to check
-        unit_name: unit to check the offsets on
     """
-    bootstrap_server = (
-        await get_address(ops_test=ops_test)
-        + f":{SECURITY_PROTOCOL_PORTS['SASL_PLAINTEXT'].client}"
-    )
+    bootstrap_servers = []
+    for unit in ops_test.model.applications[APP_NAME].units:
+        bootstrap_servers.append(
+            await get_address(ops_test=ops_test, unit_num=unit.name.split("/")[-1])
+            + f":{SECURITY_PROTOCOL_PORTS['SASL_PLAINTEXT'].client}"
+        )
 
     # example of topic offset output: 'test-topic:0:10'
     result = check_output(
-        f"JUJU_MODEL={ops_test.model_full_name} juju ssh {unit_name} sudo -i 'charmed-kafka.get-offsets --bootstrap-server {bootstrap_server} --command-config {KafkaSnap.CONF_PATH}/client.properties --topic {topic}'",
+        f"JUJU_MODEL={ops_test.model_full_name} juju ssh kafka/0 sudo -i 'charmed-kafka.get-offsets --bootstrap-server {','.join(bootstrap_servers)} --command-config {KafkaSnap.CONF_PATH}/client.properties --topic {topic}'",
```

Comment:
**question** what if `kafka/0` is down? Maybe better to use `kafka/leader`?

Reply:
Same situation. The CLI commands work without the main Kafka process being up. So if the whole unit is down, the test is not even meant to pass.

Reply:
Just a nudge here to not rely on specific unit numbers, but instead to grab one from `ops_test.model.applications[APP_NAME].units[X]`